server.ts

```diff
@@ -201,62 +212,191 @@ export function app() {
-      } else {
-        console.warn('Error in SSR, serving for direct CSR.');
-        if (hasValue(err)) {
-          console.warn('Error details : ', err);
-        }
-        res.render(indexHtml, {
-          req,
-          providers: [{
-            provide: APP_BASE_HREF,
-            useValue: req.baseUrl
-          }]
-        });
-      }
-    });
+    // Render the page to user via SSR (server side rendering)
+    serverSideRender(req, res);
   } else {
     // If preboot is disabled, just serve the client
-    console.log('Universal off, serving for direct CSR');
-    res.render(indexHtml, {
-      req,
-      providers: [{
-        provide: APP_BASE_HREF,
-        useValue: req.baseUrl
-      }]
-    });
+    console.log('Universal off, serving for direct client-side rendering (CSR)');
+    clientSideRender(req, res);
   }
 }
 
+/**
+ * Render page content on server side using Angular SSR. By default this page content is
+ * returned to the user.
+ * @param req current request
+ * @param res current response
+ * @param sendToUser if true (default), send the rendered content to the user.
+ * If false, then only save this rendered content to the in-memory cache (to refresh cache).
+ */
+function serverSideRender(req, res, sendToUser: boolean = true) {
+  // Render the page via SSR (server side rendering)
+  res.render(indexHtml, {
+    req,
+    res,
+    preboot: environment.universal.preboot,
+    async: environment.universal.async,
+    time: environment.universal.time,
+    baseUrl: environment.ui.nameSpace,
+    originUrl: environment.ui.baseUrl,
+    requestUrl: req.originalUrl,
+    providers: [{ provide: APP_BASE_HREF, useValue: req.baseUrl }]
+  }, (err, data) => {
+    if (hasNoValue(err) && hasValue(data)) {
+      res.locals.ssr = true;  // mark response as SSR (enables text compression)
+      // save server side rendered data to cache
+      saveToCache(req, data);
+      if (sendToUser) {
+        // send rendered page to user
+        res.send(data);
+      }
+    } else if (hasValue(err) && err.code === 'ERR_HTTP_HEADERS_SENT') {
+      // When this error occurs we can't fall back to CSR because the response has already been
+      // sent. These errors occur for various reasons in universal, not all of which are in our
+      // control to solve.
+      console.warn('Warning [ERR_HTTP_HEADERS_SENT]: Tried to set headers after they were sent to the client');
+    } else {
+      console.warn('Error in server-side rendering (SSR)');
+      if (hasValue(err)) {
+        console.warn('Error details : ', err);
+      }
+      if (sendToUser) {
+        console.warn('Falling back to serving direct client-side rendering (CSR).');
+        clientSideRender(req, res);
+      }
+    }
+  });
+}
+
+/**
+ * Send back response to user to trigger direct client-side rendering (CSR)
+ * @param req current request
+ * @param res current response
+ */
+function clientSideRender(req, res) {
+  res.render(indexHtml, {
```

Comment:
Calling `render()` here means that we're attempting SSR, even though it's supposed to trigger direct CSR...

I suppose this was introduced to support changing the base path in https://github.com/DSpace/dspace-angular/pull/1642, but like this we don't really have a proper fallback in case SSR fails

Reply:
Yes, that's a good point.  I also was wondering about that as I refactored this code in `server.ts` but forgot to test it out.  As you correctly noted, I didn't write this `res.render()` code, I just refactored it into a separate method.  But, we should look for a fix to this CSR fallback behavior (in a separate ticket/PR).

Reply:
Submitted an issue here: https://github.com/DSpace/dspace-angular/issues/2034